[WIP] Mitigated crash if only one font format specified, CSS merge issue

brogon opened the pull request:
If the user specified only one font format (e.g. `formats: ['woff2'],`), postcssJs creates only a single object, not a list of objects - the `.reduce()` call fails.

I've pulled the reducer function out of parseCSS and check whether it's an array or not and call `reduce()` or the reducer accordingly.

Also corrected CSS generation issue if CSS downloaded from Google Fonts only contains one `local()` definition (or more than two).  E.g. the font "Neucha" only sports one, which resulted in `srcs[2]` being undefined. So "undefined" ended up in the CSS which the browser doesn't understand and the font's whole definition is ignored.

Instead, I'm assuming the last srcs-entry (`srcs[srcs.length - 1]`) to always be the correct url() entry (this might not work/being incomplete if Google suddenly send two or more url() alternatives).

hupe1980 commented:
Well done. Only the linter errors have to be corrected. After that I would merge your PR

brogon commented:
I've corrected the style issues.

Regarding the tests: yes, I'd like to, but I'm pretty fresh to this react/ES6/node mdoules ecosystem... I'll have a look into it and orienting myself at the simple test in the __tests__ subdirectory. Let's see how far I can get - I've switched this MR to WIP for now. :)

I also found out why I had issues running the linter/jest... using yarn instead of npm helps with those 'workspaces' and missing dependencies... DX

hupe1980 commented:
No problem. Just get in touch if you need assistance. I would call the following function for the known different cases and check the result:

export async function parseCss (cssString, {fontDisplay = `swap`, useMerge})

brogon commented:
I've added tests for all four combinations (single/two locals x single/two requested formats).

I've tried two approaches: first, I put all test data in a large structure and each actual test was generated by looping over it. Then I thought those "auto generated" tests might confuse a little, so I moved everything in individually defined tests. Use whatever you prefer. :)

If you're happy with this MR, just remove the WIP & merge happily away... :)

hupe1980 commented:
Thanks 👍
